timers/timers_test.go

```diff
@@ -0,0 +1,411 @@
+package timers
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+	"unsafe"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/stretchr/testify/assert"
+)
+
+// O(n), only for use in tests.
+func (t *Timer) len() int {
+	n := 0
+	for el := t; el != nil; el = el.next {
+		n++
+	}
+	return n
+}
+
+func fakeWheel(tick time.Duration) (*Wheel, *clock.Mock) {
+	clock := clock.NewMock()
+	return newWheel(tick, 2*time.Minute, clock), clock
+}
+
+func newBucketList() *bucketList {
+	return &bucketList{
+		mask:    uint64(1<<3 - 1),
+		buckets: make([]bucket, 1<<3),
+	}
+}
+
+func assertQueuedTimers(t testing.TB, bs *bucketList, bucketIdx, expected int) {
+	b := &bs.buckets[bucketIdx]
+	b.Lock()
+	defer b.Unlock()
+	assert.Equal(
+		t,
+		expected,
+		b.timers.len(),
+		"Unexpected number of counters in bucket %v.", bucketIdx,
+	)
+}
+
+func fakeWork() {} // non-nil func to schedule
+
+func randomTimeouts(n int, max time.Duration) []time.Duration {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	ds := make([]time.Duration, n)
+	for i := 0; i < n; i++ {
+		ds[i] = time.Duration(r.Int63n(int64(max)))
+	}
+	return ds
+}
+
+func TestTimersLinkedListNilHandling(t *testing.T) {
+	// nil receiver & head
+	var root *Timer
+	assert.Equal(t, 0, root.len(), "Unexpected length for nil timer.")
+	assert.Equal(t, 0, root.push(nil).len(), "Unexpected after pushing nil on nil timer.")
+
+	// nil receiver
+	head := &Timer{}
+	root = root.push(head)
+	assert.Equal(t, 1, root.len(), "Unexpected length after pushing timer to nil root.")
+	assert.Equal(t, head, root)
+	assert.Nil(t, root.next)
+	assert.Nil(t, root.tail)
+
+	// nil head
+	originalRoot := &Timer{}
+	root = originalRoot
+	root = root.push(nil)
+	assert.Equal(t, 1, root.len(), "Unexpected length after pushing nil to a len-1 root.")
+	assert.Equal(t, originalRoot, root)
+	assert.Nil(t, root.next)
+	assert.Nil(t, root.tail)
```

Comment:
`push` should only be used with a valid head which has a `tail` pointer. `pushOne` on the other hand could work with a node that is not a head.